### PR TITLE
Don't increase commit id for read-only transactions

### PR DESCRIPTION
**Pull request by mrks**
During the work on the TPC-C, I realized that we should not commit transactions that did not change anything. There, my solution was to simply throw the transaction context away. However, it is much better to solve the underlying problem:

If a transaction does not have registered read-write operators, committing it should return immediately, leaving the global last commit ID untouched.

TPC-H (nemea, 224 cores, 100 clients):
```
+----------------+---------------+-------+---------------+-------+------------+---------------------------------+
| Benchmark      | prev. iter/s  | runs  | new iter/s    | runs  | change [%] | p-value (significant if <0.001) |
+----------------+---------------+-------+---------------+-------+------------+---------------------------------+
| TPC-H 1        | 10.0162591934 | 601   | 11.0162448883 | 661   | +10%       |                          0.0332 |
| TPC-H 2        | 128.297348022 | 7699  | 138.827362061 | 8330  | +8%        |                          0.0000 |
| TPC-H 3        | 97.7886199951 | 5868  | 101.867874146 | 6113  | +4%        |                          0.0092 |
| TPC-H 4        | 59.2916908264 | 3558  | 60.3636627197 | 3622  | +2%        |                          0.5379 |
| TPC-H 5        | 70.8966598511 | 4254  | 75.0403060913 | 4503  | +6%        |                          0.0072 |
| TPC-H 6        | 902.597839355 | 54164 | 935.067260742 | 56113 | +4%        |                          0.0000 |
| TPC-H 7        | 25.0157432556 | 1501  | 26.4642086029 | 1588  | +6%        |                          0.1348 |
| TPC-H 8        | 144.064239502 | 8645  | 153.183563232 | 9192  | +6%        |                          0.0000 |
| TPC-H 9        | 34.4781723022 | 2069  | 36.063911438  | 2164  | +5%        |                          0.1329 |
| TPC-H 10       | 66.3896865845 | 3984  | 70.2462692261 | 4215  | +6%        |                          0.0000 |
| TPC-H 11       | 500.488128662 | 30033 | 512.576049805 | 30758 | +2%        |                          0.0016 |
| TPC-H 12       | 258.119110107 | 15489 | 238.510543823 | 14312 | -8%        |                          0.0000 |
| TPC-H 13       | 51.908405304  | 3115  | 52.7261276245 | 3164  | +2%        |                          0.3952 |
| TPC-H 14       | 620.860046387 | 37252 | 597.569335938 | 35856 | -4%        |                          0.0000 |
| TPC-H 15       | 504.935119629 | 30301 | 469.191497803 | 28152 | -7%        |                          0.0000 |
| TPC-H 16       | 180.371154785 | 10823 | 178.472457886 | 10709 | -1%        |                          0.1971 |
| TPC-H 17       | 24.41264534   | 1465  | 24.2799701691 | 1457  | -1%        |                          0.7094 |
| TPC-H 18       | 25.4134922028 | 1525  | 25.4625873566 | 1528  | +0%        |                          0.6827 |
| TPC-H 19       | 131.058868408 | 7864  | 128.728607178 | 7724  | -2%        |                          0.2459 |
| TPC-H 20       | 46.94764328   | 2817  | 46.1290588379 | 2768  | -2%        |                          0.3491 |
| TPC-H 21       | 19.9805049896 | 1199  | 19.849029541  | 1191  | -1%        |                          0.6078 |
| TPC-H 22       | 186.092056274 | 11167 | 185.744537354 | 11145 | -0%        |                          0.7061 |
| geometric mean |               |       |               |       | +2%        |                                 |
+----------------+---------------+-------+---------------+-------+------------+---------------------------------+
```

The machine was not 100% empty, so the numbers are not perfect. As this should not break anything, I would say that they are good enough.

### CODE REVIEW

File: src/test/concurrency/transaction_context_test.cpp [hunk 1]
```diff
@@ -11,14 +11,25 @@
...
+#include "operators/get_table.hpp"
+#include "operators/validate.hpp"
 #include "types.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
 class TransactionContextTest : public BaseTest {
+ public:
+  std::string table_name;
```
Comment (julianmenzler) [inline src/test/concurrency/transaction_context_test.cpp]: What do you think about the following?
```suggestion
  static constexpr auto table_name = "test_table";
```

File: src/test/concurrency/transaction_context_test.cpp [hunk 2]
```diff
@@ -11,14 +11,25 @@
...
  protected:
-  void SetUp() override {}
+  void SetUp() override {
+    table_name = "test_table";
+    auto t = load_table("resources/test_data/tbl/float_int.tbl");
+    // Insert Operator works with the Storage Manager, so the test table must also be known to the StorageManager
+    StorageManager::get().add_table(table_name, t);
+  }
 
   TransactionManager& manager() { return TransactionManager::get(); }
```
Comment (julianmenzler) [inline src/test/concurrency/transaction_context_test.cpp]: Adopt the new Hyrise singleton approach.

File: src/lib/concurrency/transaction_context.cpp [hunk 3]
```diff
@@ -88,6 +88,12 @@ void TransactionContext::commit_async(const std::function<void(TransactionID)>&
 void TransactionContext::commit() {
   Assert(_phase == TransactionPhase::Active, "TransactionContext must be active to be committed.");
 
+  // No modifications made, nothing to commit, no need to acquire a commit ID
+  if (_rw_operators.empty()) {
+    _phase = TransactionPhase::Committed;
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/concurrency/transaction_context.cpp]: Good point. Forgot about that.

**Comment by mrks** [comment]
Great suggestions; thanks for looking at the bigger picture, too.

> Why don't we calll
register_read_write_operator(std::shared_ptr<AbstractReadWriteOperator> op) in AbstractReadWriteOperator::execute()?
Currently, we call it in each ReadWriteOperator separately, Insert and Delete.

Done
Good idea.

**Comment by julianmenzler** [comment]
I'll have a look later this evening.

**Pull request merged**